src/index.d.ts

```diff
@@ -0,0 +1,120 @@
+declare module "hyperapp" {
+  // A Hyperapp application instance has an initial state and a base view.
+  // It must also be mounted over an available DOM element.
+  type App<S, P, D> = Readonly<{
+    init: Action<S, P, D>;
+    view: View;
+    node: Node;
+    subscriptions?: Subscription;
+    middleware?: Middleware;
+  }>
+
+  // A view builds a virtual DOM node representation of the application state.
+  type View = <S>(state: State<S>) => VDOM
+
+  // Application state is accessible in every view, action, and subscription.
+  type State<S> = S
+
+  // A subscription is a set of recurring effects.
+  type Subscription = <S>(state: State<S>) => Subscriber[]
+
+  // A subscriber reacts to subscription updates.
+  type Subscriber = boolean | void | Effect | Unsubscribe
+
+  // A subscriber ideally provides a function that cancels itself properly.
+  type Unsubscribe = () => void
+
+  // Middleware allows for custom processing during dispatching.
+  type Middleware = (f: Dispatch) => Dispatch
+
+  // -----------------------------------------------------------------------------
+
+  // A dispatch is a response to an event within the context of the current state.
+  type Dispatch = <S, P, D>(action: Action<S, P, D>, props?: Payload<P>) => void
+
+  // An action transforms existing state while possibly invoking effects.
+  type Action<S, P, D>
+    // An action can wrap another action.
+    = ((state: State<S>, props?: Payload<P>) => Action<S, P, D>)
+    // An action can be given a payload.
+    | [Action<S, P, D>, Payload<P>]
+    // State can be set directly along with any effects to run.
+    // This is particularly when setting the initial state.
+    | [State<S>, ...EffectDescriptor<D>[]]
+    | State<S>
+
+  // A payload allows data not within the state to be given to a dispatch.
+  type Payload<P> = P
+
+  // An effect descriptor describes how an effect should be invoked.
+  // A function that creates this is called an effect constructor.
+  type EffectDescriptor<D> = [Effect, EffectData<D>]
+
+  // An effect is where side effects and any additional dispatching occur.
+  type Effect = <D>(dispatch: Dispatch, props?: EffectData<D>) => void
+
+  // An effect is generally given additional data.
+  type EffectData<D> = D
+
+  // -----------------------------------------------------------------------------
+
+  // A virtual DOM node represents an actual DOM element.
+  type VDOM = {
+    readonly type: string;
+    readonly props: PropList;
+    readonly children: VNode[];
+    node: MaybeNode;
+    readonly tag?: Tag;
+    readonly key: Key;
+    memo?: PropList;
+  }
+
+  // Virtual DOM properties will often correspond to HTML attributes.
+  type PropList = Record<string, Prop> & Readonly<{ key?: Key }> & ElementCreationOptions
+  type Prop = bigint | boolean | null | number | string | symbol | undefined | Function | ClassProp | StyleProp
```

Comment:
Couldn't there also be Arrays or Object of any signature be a prop value ? For example `{foo: string: bar: number}[]`, and shouldn't the component dictate what explicit type the property has ?

Reply:
So, `Prop` exists primarily to describe the properties `h` would be able to work with. As far as I'm aware the only prop that can have an array value or object value is `class` while `style` must have an object value.

I'm hesitant to broaden the scope of `Prop` to let it be directly relied upon in user-created components. I want `Prop` to maintain a level of precision that doesn't make it a good fit for that purpose. It would be overkill in some ways and too restrictive in others.

Instead, I encourage creating component-level prop types tailored to your components.